core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java

```diff
@@ -301,40 +314,59 @@ private NodeInfo asNodeInfo(AdminRow row) {
     return builder;
   }
 
-  private Optional<NodeInfo> findInPeers(AdminResult result, InetSocketAddress connectAddress) {
-    // The peers table is keyed by broadcast_address, but we only have the translated
-    // broadcast_rpc_address, so we have to traverse the whole table and check the rows one by one.
+  private Optional<NodeInfo> findInPeers(
+      AdminResult result, InetSocketAddress broadcastRpcAddressToFind) {
+    // The peers table is keyed by broadcast_address, but we only have the broadcast_rpc_address, so
+    // we have to traverse the whole table and check the rows one by one.
+    for (AdminRow row : result) {
+      InetSocketAddress broadcastRpcAddress = getBroadcastRpcAddress(row);
+      if (broadcastRpcAddress != null && broadcastRpcAddress.equals(broadcastRpcAddressToFind)) {
+        return Optional.of(nodeInfoBuilder(row, broadcastRpcAddress).build());
+      }
+    }
+    LOG.debug("[{}] Could not find any peer row matching {}", logPrefix, broadcastRpcAddressToFind);
+    return Optional.empty();
+  }
+
+  private Optional<NodeInfo> findInPeers(AdminResult result, UUID hostIdToFind) {
     for (AdminRow row : result) {
-      InetSocketAddress nativeAddress = getNativeAddressFromPeers(row);
-      if (nativeAddress != null && nativeAddress.equals(connectAddress)) {
-        return Optional.of(nodeInfoBuilder(row, nativeAddress).build());
+      UUID hostId = row.getUuid("host_id");
+      if (hostId != null && hostId.equals(hostIdToFind)) {
+        return Optional.of(nodeInfoBuilder(row, getBroadcastRpcAddress(row)).build());
       }
     }
-    LOG.debug("[{}] Could not find any peer row matching {}", logPrefix, connectAddress);
+    LOG.debug("[{}] Could not find any peer row matching {}", logPrefix, hostIdToFind);
     return Optional.empty();
   }
 
   // Current versions of Cassandra (3.11 at the time of writing), require the same port for all
   // nodes. As a consequence, the port is not stored in system tables.
   // We save it the first time we get a control connection channel.
   private void savePort(DriverChannel channel) {
-    if (port < 0 && channel.connectAddress() instanceof InetSocketAddress) {
-      port = ((InetSocketAddress) channel.connectAddress()).getPort();
+    if (port < 0) {
+      SocketAddress address = channel.getEndPoint().resolve();
+      if (address instanceof InetSocketAddress) {
+        port = ((InetSocketAddress) address).getPort();
+      }
     }
   }
 
-  private InetSocketAddress getNativeAddressFromPeers(AdminRow row) {
+  private InetSocketAddress getBroadcastRpcAddress(AdminRow row) {
```

Comment:
Shouldn't this method be on the `AdminRow` class?

Reply:
`AdminRow` is a simplified row implementation used by internal admin tasks, but it's still a generic utility that can be used with any table.
This method is used specifically for `system.local` and `system.peers`, so I feel that it doesn't belong to the generic type.